### PR TITLE
Add Dramatiq instrumentation

### LINKED ISSUE #998: support for dramatiq tasks
### Description

Mainly the title, but was wondering if this was on the roadmap

### PR DESCRIPTION
**Pull request by dmontagu**
Closes #998.

## What changed

- Add `logfire.instrument_dramatiq()` and a first-party Dramatiq middleware.
- Propagate creation and delivery context through messages and link retries/requeues to their originating work.
- Model enqueue and worker delivery lifecycles without leaking OpenTelemetry context across worker threads.
- Handle success, failure, retry, skip, requeue, nack, delayed messages, concurrent deliveries, teardown, and reinstrumentation.
- Add the public integration module, no-op shim, optional dependency, documentation, and navigation.

## Why

Dramatiq users could assemble OpenTelemetry behavior manually, but correct message propagation and worker lifecycle handling is subtle—especially around retries and thread-local context.

## Validation

- Focused integration and shim suite: 18 passed, 1 skipped.
- 15 focused lifecycle tests passed on Dramatiq 1.13, 1.18, 2.1, and 2.2.
- Documentation tests.
- Ruff, Ruff format, Pyright, `uv lock --check`, pre-commit hooks, and `git diff --check`.

<a href="https://cubic.dev/pr/pydantic/logfire/pull/2250?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: docs/integrations/event-streams/dramatiq.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+logfire.instrument_dramatiq(broker)
+dramatiq.set_broker(broker)
+
+
+@dramatiq.actor
+def send_email(address: str) -> None:
+    print(f'Email sent to {address}')
+
+
+send_email.send('hello@example.com')
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/integrations/event-streams/dramatiq.md]: Fixed in e401593b: the example enqueue is now guarded by if __name__ == '__main__', so importing the actor module from the Dramatiq CLI does not send a message.

File: logfire/_internal/integrations/dramatiq.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    ) -> None:
+        self._finish_message(message, exception)
+
+    def after_skip_message(self, broker: Broker, message: MessageProxy) -> None:
+        self._finish_message(message, None)
+
+    def after_nack(self, broker: Broker, message: MessageProxy) -> None:
+        self._finish_message(message, None)
+
+    def after_worker_shutdown(self, broker: Broker, worker: Any) -> None:
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/dramatiq.py:133]: Fixed in e401593b: after_worker_shutdown no longer calls broker-wide close(). Normal terminal hooks own each delivery, while explicit close()/uninstrument() handle intentional global cleanup. Added a concurrent/shared-broker regression test.

File: logfire/_internal/integrations/dramatiq.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            f'{message.actor_name} process',
+            _span_kind=SpanKind.CONSUMER,
+            _links=links,
+            **_attributes(message, 'process'),
+        )
+        # Starting under the delivery context establishes the parent without making the consumer
+        # span current. Dramatiq may finish or cancel work from another thread, where detaching an
+        # OpenTelemetry context token created here would be invalid.
+        with logfire_propagate.attach_context(delivery):
+            span._start()  # pyright: ignore[reportPrivateUsage]
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/dramatiq.py:117]: Fixed in e401593b: the delivery span is attached on the worker thread for actor execution and detached on that same thread during normal terminal lifecycle. Child telemetry now nests under the consumer span; cross-thread close remains safe. Added regression coverage.

File: logfire/_internal/integrations/dramatiq.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        self._finish_message(message, exception)
+
+    def after_skip_message(self, broker: Broker, message: MessageProxy) -> None:
+        self._finish_message(message, None)
+
+    def after_nack(self, broker: Broker, message: MessageProxy) -> None:
+        self._finish_message(message, None)
+
+    def after_worker_shutdown(self, broker: Broker, worker: Any) -> None:
+        self.close()
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/dramatiq.py]: Fixed in e401593b: worker shutdown no longer performs broker-wide delivery cleanup. Terminal message hooks finish their own spans, while explicit close()/uninstrument() retain global teardown ownership. Added a shared-broker regression test.

File: docs/integrations/event-streams/dramatiq.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/integrations/event-streams/dramatiq.md]: Fixed in e401593b: send_email.send(...) is under the module's __main__ guard, preventing an enqueue when dramatiq imports the task module.

File: logfire/_internal/integrations/dramatiq.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        result = json.loads(value)
+    except (TypeError, ValueError):
+        return {}
+    if not isinstance(result, dict) or not all(isinstance(k, str) and isinstance(v, str) for k, v in result.items()):
+        return {}
+    return result
+
+
+def _span_context(carrier: dict[str, str]) -> trace.SpanContext | None:
+    with logfire_propagate.attach_context(carrier):
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/dramatiq.py]: Fixed in e401593b: all creation and delivery carrier extraction paths now pass third_party=True. Added a regression proving crafted traceparent values are ignored when distributed tracing is disabled.

File: logfire/_internal/integrations/dramatiq.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    import dramatiq
+    from dramatiq import Broker, Message
+    from dramatiq.broker import MessageProxy
+    from dramatiq.middleware import Middleware, Retries
+except ImportError as e:  # pragma: no cover
+    raise RuntimeError(
+        '`logfire.instrument_dramatiq()` requires the `dramatiq` package.\n'
+        'You can install this with:\n'
+        "    pip install 'logfire[dramatiq]'"
+    ) from e
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/dramatiq.py]: Fixed in e401593b: the optional dependency guard now catches ModuleNotFoundError and only translates the error when e.name == 'dramatiq'; unrelated missing dependencies are re-raised. Added a regression test.

File: logfire/_internal/integrations/dramatiq.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+    try:
+        result = json.loads(value)
+    except (TypeError, ValueError):
+        return {}
+    if not isinstance(result, dict) or not all(isinstance(k, str) and isinstance(v, str) for k, v in result.items()):
+        return {}
+    return result
+
+
+def _span_context(carrier: dict[str, str]) -> trace.SpanContext | None:
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply logfire/_internal/integrations/dramatiq.py:61]: Fixed in e401593b: all decoded message carriers are treated as third-party context. The new distributed_tracing=False regression covers both creation and delivery carriers.

File: logfire/_internal/integrations/dramatiq.py
Comment (alexmojaki) [inline logfire/_internal/integrations/dramatiq.py:1]: why roll our own instead of using https://pypi.org/project/opentelemetry-instrumentor-dramatiq/ as in the issue?

Comment (dmontagu) [reply logfire/_internal/integrations/dramatiq.py:1]: Good question. I re-tested the latest release, 0.1.4, rather than relying on the earlier prototype. It is a third-party instrumentor rather than an OpenTelemetry contrib package, and it is not currently safe for Logfire to wrap: standard `message_with_options(delay=..., max_retries=...)` values are nested under an `options` key, so Dramatiq does not see them; producer spans use `SERVER` instead of `PRODUCER`; actor code sees the producer context rather than the consumer span because the consumer is not made current; extracted context is attached but never detached; and it monkey-patches the private `_WorkerThread` implementation. Its test suite only checks message tuple behavior and does not run a worker, retry, delay, or failure. The implementation here uses Dramatiq middleware and covers versions 1.13 through 2.2, retries, delays, concurrency, failures, and cleanup. That is why I did not delegate to the package. If you still prefer a thin wrapper despite those behavior gaps, I can rewrite it that way.

Comment (alexmojaki) [reply logfire/_internal/integrations/dramatiq.py:1]: elaborate and explain this, we know nothing about dramatiq

Comment (dmontagu) [reply logfire/_internal/integrations/dramatiq.py:1]: Here is the missing Dramatiq model and why these details matter.

Dramatiq is a background-job queue, similar in shape to Celery. Application code calls an actor's `send()` method, which creates a message and puts it on a broker. A worker thread later takes that message from the broker and calls the actor. Dramatiq's retry middleware can put a failed message back on the broker for another delivery.

Correct tracing therefore needs two operations:

1. A producer span around enqueueing the message, with the current trace context stored on the message.
2. A consumer span around the worker delivery, made current while the actor runs, then reliably removed from that reusable worker thread.

The current third-party 0.1.4 implementation has problems in both Dramatiq behavior and trace behavior:

- It calls Dramatiq's `message_with_options(..., options={...})`, even though Dramatiq receives options through `**options`. This puts all options inside `message.options['options']` instead of at the top level where Dramatiq middleware reads values such as retry configuration. Its only test explicitly asserts that nested layout. This can change job behavior, not merely span presentation.
- It labels the enqueue operation as an OpenTelemetry `SERVER` span. Enqueueing is a `PRODUCER` operation; `SERVER` means handling an inbound request and gives Logfire and OTel consumers the wrong semantics.
- The worker uses `start_span()`, which creates a consumer span without making it current. It instead makes the extracted producer context current. Spans created inside the actor therefore become children of the enqueue span rather than the worker-processing span, so the trace does not show what happened during message processing.
- It calls `context.attach()` on the worker thread but never calls `context.detach()`. Dramatiq reuses worker threads, so a later unrelated message can inherit an earlier message's trace context.
- It monkey-patches Dramatiq's private `_WorkerThread.process_message`. That is not a supported extension point and increases compatibility risk across Dramatiq versions.
- The package is third-party rather than part of OpenTelemetry contrib, and its test suite has one message-construction test. It does not run a worker or cover retry, delay, failure, concurrent delivery, or shutdown behavior.

The exact implementation reviewed is [here](https://github.com/Shrest4647/opentelemetry-instrumentor-dramatiq/blob/4dc0cd72060e14f804170a550f4abba3cd7827a3/opentelemetry_instrumentor_dramatiq/src/__init__.py), and its test is [here](https://github.com/Shrest4647/opentelemetry-instrumentor-dramatiq/blob/4dc0cd72060e14f804170a550f4abba3cd7827a3/tests/test_instrumentation.py).

This PR instead uses Dramatiq's public middleware hooks for worker lifecycle and wraps broker enqueueing for producer spans. The extra code handles retries, delayed messages, skip/nack paths, concurrent workers, context cleanup, shutdown, and uninstrumentation, with tests against Dramatiq 1.13, 1.18, 2.1, and 2.2.

The tradeoff is that we would own a real Dramatiq integration, not a tiny convenience wrapper. My recommendation is: if we do not want that maintenance responsibility because the team lacks Dramatiq expertise, park or close this PR and fix the third-party instrumentor upstream first. If we do want first-party Dramatiq support, this implementation is safer than wrapping 0.1.4. I do not recommend exposing 0.1.4 through `logfire.instrument_dramatiq()` as it stands.
